### PR TITLE
Moodle Client Updates

### PR DESCRIPTION
**Pull request by jlajoie**
**Merge checklist:**
- [ ] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files)
    - `base.in` if needed in production but edx-platform doesn't install it
    - `test-master.in` if edx-platform pins it, with a matching version
    - `make upgrade && make requirements` have been run to regenerate requirements
- [ ] `make static` has been run to update webpack bundling if any static content was updated
- [ ] `./manage.py makemigrations` has been run
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
    - This should be run from either a venv with all the lms/edx-enterprise requirements installed or if you checked out edx-enterprise into the src directory used by lms, you can run this command through an lms shell.
        - It would be `./manage.py lms makemigrations` in the shell.
- [ ] [Version](https://github.com/edx/edx-enterprise/blob/master/enterprise/__init__.py) bumped
- [ ] [Changelog](https://github.com/edx/edx-enterprise/blob/master/CHANGELOG.rst) record added
- [ ] Translations updated (see docs/internationalization.rst but also this isn't blocking for merge atm)

**Post merge:**
- [ ] Tag pushed and a new [version](https://github.com/edx/edx-enterprise/releases) released
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
- [ ] After versioned build finishes in [Travis](https://travis-ci.org/github/edx/edx-enterprise), verify version has been pushed to [PyPI](https://pypi.org/project/edx-enterprise/)
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
- [ ] PR created in [edx-platform](https://github.com/edx/edx-platform) to upgrade dependencies (including edx-enterprise)
    - This **must** be done after the version is visible in PyPi as `make upgrade` in edx-platform will look for the latest version in PyPi.
    - Note: the edx-enterprise constraint in edx-platform **must** also be bumped to the latest version in PyPi.

### CODE REVIEW

File: integrated_channels/moodle/client.py
Comment (taliaedX) [inline integrated_channels/moodle/client.py:8]: hmmm why this specific import path instead of just `from urllib.parse`? Especially curious since there is an import from urllib.parse already in this file (I imported urlencode because I wanted the base functionality not Django's wrapped version)

File: integrated_channels/moodle/client.py
Comment (taliaedX) [inline integrated_channels/moodle/client.py:54]: Hmmm. Given the tokens seem to persist on the moodle side, _should_ we be getting a new token every time instead of persisting it? Would we not end up possibly hitting some data limit or just otherwise polluting the web service token screen with 8000 tokens? lol

Comment (taliaedX) [reply integrated_channels/moodle/client.py:54]: secondly: mostly a note for myself, but if we go this route, I need to update my parameters as it's using `self.enterprise_configuration.api_token` in several spots.

File: integrated_channels/moodle/client.py
Comment (taliaedX) [inline integrated_channels/moodle/client.py:142]: Is this only required for token requests? Mostly just curious since I have never included it, but since tokens seem to be assigned per webservice that might make sense.

File: integrated_channels/moodle/client.py
Comment (jlajoie) [reply integrated_channels/moodle/client.py:8]: Bad merge and shamelessly stolen from elsewhere, I'll update and simplify

File: integrated_channels/moodle/client.py
Comment (jlajoie) [reply integrated_channels/moodle/client.py:54]: So the way I've seen it is it doesn't actually create a new token every time, if you spam this auth endpoint as the same user you'll get the same token returned.

Let me hold off on merging this, I'll see what it's like with a TTL set on token expiry just to verify that's true in that scenario as well, good catch here.

File: integrated_channels/moodle/client.py
Comment (jlajoie) [reply integrated_channels/moodle/client.py:142]: Yup! This'll give you the token for a service with specific functions/permissions. That token is then used in Moodle elsewhere (and I'm guessing they resolve the token to a service to get perms) so you don't need to specify that name in other requests.

**Comment by jlajoie** [comment]
Also realizing we should include the categoryid in here as well, going to update migrations

File: integrated_channels/moodle/client.py
Comment (taliaedX) [reply integrated_channels/moodle/client.py:54]: oh yeah I forgot it returns the same token. I had that happen too when we tried testing this a while back.

**Comment by taliaedX** [comment]
>Also realizing we should include the categoryid in here as well, going to update migrations

Hmmm. Should we? Wouldn't the default be the same for every customer? Or did you check with Ken about custom categories?

File: integrated_channels/moodle/client.py
Comment (taliaedX) [reply integrated_channels/moodle/client.py:54]: I also wonder if this wrapper could just pass token or username/password and the base url in the format of `base_params = {'moodlewsrestformat': 'json', 'wstoken': self.token, 'wsusername', [etc]}` since you might see I have to use those same values in every single request, but that's something I can also play around with outside of this particular pr.

File: integrated_channels/moodle/migrations/0001_initial.py
Comment (taliaedX) [inline integrated_channels/moodle/migrations/0001_initial.py:1]: look at you using an old version of Django. :p

Comment (jlajoie) [reply integrated_channels/moodle/migrations/0001_initial.py:1]: This is with a brand new install in my venv too!

**Comment by jlajoie** [comment]
@taliaedX going to do some smoke testing with everything you have up then merge this into your branch if things look good

**Comment by taliaedX** [comment]
I'm working on fixing tests, but @alex-sheehan-edx I'd love a review whenever you have time and @jlajoie I know it's your PR, but I'd appreciate you looking over the latest commit changes to double-check my changes.

**Comment by taliaedX** [comment]
Couple side notes:
1. I'm already planning on cleaning up any commented out bits, unnecessary comments, print statements, etc. as part of style fixes.
2. None of the merge checklist is relevant here since it's being merged into my PR not master.

File: integrated_channels/moodle/client.py
Comment (jlajoie) [inline integrated_channels/moodle/client.py:103]: Might want to add some handling around array length and key existence. You'll also not want the moodle_request_wrapper on this if you're not returning a raw response

**Review by jlajoie — COMMENTED** [review]
Just the one comment but looks good

**Review by alex-sheehan-edx — APPROVED** [review]
Took a high level first pass. Looks pretty good to me, but talked with @taliaedX and will probably do a more in-depth review of the main PR for moodle work.

File: integrated_channels/moodle/client.py
Comment (taliaedX) [reply integrated_channels/moodle/client.py:119]: ah I missed that one. Good catch

File: integrated_channels/moodle/models.py
Comment (taliaedX) [reply integrated_channels/moodle/models.py:48]: I named them wsusername and wspassword since those are the actual parameter names in Moodle's api.
My guess is Jeff changed them since these field names would be visible in Django admin and might confuse support or something idk.

@jlajoie why you hatin' on my field names?

**Pull request merged**
